core/src/test/java/org/infinispan/api/BaseCacheAPIPessimisticTest.java

```diff
@@ -0,0 +1,156 @@
+
+      // Wait until fork thread has alredy locked the key
+      barrier.await(10, TimeUnit.SECONDS);
+
+      LockedStream<Object, Object> stream = cache.getAdvancedCache().lockedStream();
+      Future<?> forEachFuture = fork(() -> stream.filter(e -> e.getKey().equals(key)).forEach((c, e) ->
+            assertEquals("value" + key + "-new", c.put(e.getKey(), String.valueOf(e.getValue() + "-other")))));
+
+      try {
+         forEachFuture.get(50, TimeUnit.MILLISECONDS);
+         fail("Test should have thrown TimeoutException");
+      } catch (TimeoutException e) {
+
+      }
+
+      // Let the tx put complete
+      barrier.await(10, TimeUnit.SECONDS);
+
+      forEachFuture.get(10, TimeUnit.MINUTES);
+
+      // The put should replace the value that forEach inserted
+      assertEquals("value" + key, putFuture.get(10, TimeUnit.SECONDS));
+      // The put should be last since it had to wait until lock was released on forEachWithLock
+      assertEquals("value" + key + "-new-other", cache.get(key));
+
+      // Make sure the locks were cleaned up properly
+      LockManager lockManager = cache.getAdvancedCache().getComponentRegistry().getComponent(LockManager.class);
+      assertEquals(0, lockManager.getNumberOfLocksHeld());
+   }
+
+   @DataProvider(name = "testLockedStreamInTx")
+   public Object[][] testLockedStreamInTxProvider() {
+      return new Object[][] { { Boolean.TRUE }, { Boolean.FALSE} };
+   }
+
+   @Test(dataProvider = "testLockedStreamInTx")
+   public void testLockedStreamInTxCommit(Boolean shouldCommit) throws SystemException, NotSupportedException,
+         HeuristicRollbackException, HeuristicMixedException, RollbackException {
+      for (int i = 0; i < 5; i++) {
+         cache.put(i, "value" + i);
+      }
+
+      TransactionManager tm = cache.getAdvancedCache().getTransactionManager();
+      tm.begin();
+      try {
+         cache.getAdvancedCache().lockedStream().forEach((c, e) -> c.put(e.getKey(), e.getValue() + "-changed"));
+         if (shouldCommit) {
+            tm.commit();
+         } else {
+            tm.rollback();
+         }
+      } finally {
+         if (tm.getStatus() == Status.STATUS_ACTIVE) {
+            tm.rollback();
+         }
+      }
+
+      for (int i = 0; i < 5; i++) {
+         assertEquals("value" + i + "-changed", cache.get(i));
+      }
+   }
+
+   public void testLockedStreamTxInsideConsumer() {
+      for (int i = 0; i < 5; i++) {
+         cache.put(i, "value" + i);
+      }
+
+      cache.getAdvancedCache().lockedStream().forEach((c, e) -> {
+         TransactionManager tm = c.getAdvancedCache().getTransactionManager();
+         try {
+            tm.begin();
+            c.put(e.getKey(), e.getValue() + "-changed");
+            tm.commit();
+         } catch (Exception e1) {
+            throw new RuntimeException(e1);
+         }
+      });
+
+      for (int i = 0; i < 5; i++) {
+         assertEquals("value" + i + "-changed", cache.get(i));
+      }
+   }
```

Comment:
I still miss the case when you have both outer and inner transaction (though from the code I am pretty confident it will work at this moment).